src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private void linkFirst(T value) {
+        final Node<T> h = head;
+        final Node<T> newNode = new Node<>(null, value, h);
+        head = newNode;
+        if (h == null) {
+            tail = newNode;
+        } else {
+            h.prev = newNode;
+        }
+        size++;
+    }
+
+    private void linkLast(T value) {
+        final Node<T> last = tail;
+        final Node<T> newNode = new Node<>(last, value, null);
+        tail = newNode;
+        if (last == null) {
+            head = newNode;
+        } else {
+            last.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T t, Node<T> beOnTheRight) {
+        final Node<T> prevActual = beOnTheRight.prev;
+        final Node<T> actual = new Node<>(prevActual, t, beOnTheRight);
+        beOnTheRight.prev = actual;
+        if (prevActual == null) {
+            head = actual;
+        } else {
+            prevActual.next = actual;
+        }
+        size++;
+    }
+
+    private Node<T> searchByIndex(int index) {
+        Node<T> search;
+        if (index < (size >> 1)) {
+            search = head;
+            for (int i = 0; i < index; i++) {
+                search = search.next;
+            }
+        } else {
+            search = tail;
+            for (int i = size - 1; i > index; i--) {
+                search = search.prev;
+            }
+        }
+        return search;
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            linkFirst(value);
+        } else {
+            linkLast(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index Out Of Bound");
+        }
```

Comment:
Extract to separate method

Reply:
I have almost the same method. the only difference is ">=". Therefore, I decided not to put it in a separate method, since it is used only once